staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go

```diff
@@ -282,6 +282,11 @@ func (h *UpgradeAwareHandler) tryUpgrade(w http.ResponseWriter, req *http.Reques
 		location = *req.URL
 		location.Scheme = h.Location.Scheme
 		location.Host = h.Location.Host
+	} else if len(location.RawQuery) == 0 {
+		// When a connection is upgraded, we want to keep the request querystring, even if the request location
+		// is not being used. However, if the location already has its own querystring, we should not overwrite
```

Comment:
I think it is a totally valid concern, and I would be OK with closing this PR and consider it a known limitation of proxy, that you just can't pass query string parameters on a proxied websocket request.  

Besides, this PR has been around over a year and hasn't really gotten a lot of interest.  Personally, I don't have a stake in whether it gets fixed or not, I just reproduced the issue and worked on a fix that I thought would work.

It certainly seems like an edge case I'll admit, and given that, I'm not sure it is worth the potential risk even though I don't immediately see a risk of adding a query string to the request.  As with all things security-related, you can be 99% right and 1% wrong and its the 1% that matters, so I get it.

I guess a workaround might be to [use `kubectl port-forward` instead](https://github.com/kubernetes/kubernetes/issues/89360#issuecomment-797207790)

Reply:
it looks like for _non-upgrade_ requests, we use the query from the request unconditionally:

https://github.com/kubernetes/kubernetes/blob/14df10f60bbd68216f8ab9d3f30d1f938d7b659e/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L203-L204

to help decide whether this change is safe or needs to be limited to specific instances (like the node/pod/service proxy endpoints), can the following info be gathered?

1. enumerate the places we use `NewUpgradeAwareHandler` in the apiserver and kubelet
2. of those, determine which uses don't already honor the request query (by not setting `UseRequestLocation` to true, and by not explicitly honoring the request query like `staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go` does)
3. of those, determine which uses are targeting backends that could support upgrade requests
4. of those, determine which uses are not populating the backend query themselves 100% of the time
5. of those, determine which uses have known backend parameters that would be problematic to allow control over (the exec/attach endpoints especially come to mind, if the user could send a POST to the API server exec/attach endpoints with both a query and a body, get the body to be used for admission/audit inspection, then override the parameters with a query string)

Reply:
Thanks, @liggitt 

I've looked through the code and found the places where `NewUpgradeAwareHandler` is being used:
* pod exec/attach/portforward/proxy (`pkg/registry/core/pod/rest/subresource.go`)
* service proxy (`pkg/registry/core/service/proxy.go`)
* node proxy (`pkg/registry/core/node/rest/proxy.go`)
* kube-aggregator (`staging/src/k8s.io/kube-aggregator/pkg/apiserver`)
* kubectl proxy (`staging/src/k8s.io/kubectl/pkg/proxy`)

Of those, it appears that only kubectl proxy sets `UseRequestLocation` to true.

I'm not quite sure how to tell which of those are targeting backends that could support upgrade requests.  So that's where I left off.

---

However, I've been thinking it over and **want to get your thoughts on a different approach**:

What if instead of automatically applying the request query like this PR is doing, we add another option to `UpgradeAwareHandler`, called something like`UseRequestQuery`, which works in a way similar to `UseRequestLocation`?  

This option would make it an explicit choice to use the request query without the request location.

The advantage I see to this is that it would be possible to make this new behavior apply *only* to proxy, leaving exec, portforward, and attach to work the same as they do today.

I opened a draft PR to show an example of what I'm talking about. It still probably needs some work, but I created this to illustrate the idea:
https://github.com/kubernetes/kubernetes/pull/102002/files

Reply:
requiring opt-in is certainly easier to reason about... I like the direction of that PR

would want to pick a good name, and think about / document well how it interacts with the other options on the proxier

Reply:
I could go either way really.

Opt-in has the advantage of limiting the scope of the change, but it also seems to add some additional complexity (yet another flag).  While it is a safer change, it may not really be the best change from a code perspective due to the already somewhat large number of flags that affect how it works.

To add a point about the original approach of just always preserving the request query:

Doing that for upgraded requests would be consistent with what it does currently in the non-upgrade scenario:
https://github.com/kubernetes/kubernetes/blob/3d82e8e3c1eef729df36c56c97fc461c26a97738/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L203-L204

To simplify this PR as much as possible, I've updated my commit to just do exactly the same thing in the upgrade scenario as the non-upgrade scenario:
https://github.com/kubernetes/kubernetes/blob/3d82e8e3c1eef729df36c56c97fc461c26a97738/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L280-L281

This would bring into alignment how it handles the query in both the **upgrade** and **non-upgrade** scenarios so they behave the same in both cases.

Your previous points about how to determine whether this is safe in all cases are still valid and I will need to take some time to investigate further.